src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/AzureSqlKeyCryptographer.cs

```diff
@@ -52,18 +50,34 @@ internal AzureSqlKeyCryptographer(TokenCredential tokenCredential)
 
         /// <summary>
         /// Adds the key, specified by the Key Identifier URI, to the cache.
+        /// Validates the key type and fetches the key from Azure Key Vault if it is not already cached.
         /// </summary>
         /// <param name="keyIdentifierUri"></param>
         internal void AddKey(string keyIdentifierUri)
         {
-            if (TheKeyHasNotBeenCached(keyIdentifierUri))
+            try
             {
-                ParseAKVPath(keyIdentifierUri, out Uri vaultUri, out string keyName, out string keyVersion);
-                CreateKeyClient(vaultUri);
-                FetchKey(vaultUri, keyName, keyVersion, keyIdentifierUri);
-            }
+                // Allow only one thread to proceed to ensure thread safety
+                // as we will need to fetch key information from Azure Key Vault if the key is not found in cache.
+                _keyDictionarySemaphore.Wait();
```

Comment:
Move `_keyDictionarySemaphore.Wait();` to before the try. We don't want to execute the finally if Wait() throws an exception.